### ISSUE TITLE
Restic 0.9.0 + Windows 7 cmd.exe + CygWin64 Bash : interoperability issues

### ISSUE THREAD

**Issue by sdrubble** [restic/restic#1838]
## Output of `restic version` 

restic 0.9.0 compiled with go1.10.2 on windows/amd64

## How did you run restic exactly?

A Windows .bat file (which runs under Windows 7 'cmd.exe') and calls CigWin64's bash.exe , which in turn runs a 730-line BASH script with a number of external BASH subroutines.

The .bat file, the main script and all the subroutines manipulate a number of environment variables.

Previously I had successfully used a similar combination to use restic commands 'check' , 'ls' , and 'snapshots', while also generating logs.

I'm having a number of issues when implementing a script to run restic's 'backup' command, all of them related to specifying a directory name from within a variable. While backing up directories WITHOUT the _space_ character in their pathnames has worked as expected, I have been unable to make restic accept pathnames WITH spaces, regardless of single quotes, double quotes, or assorted alternatives with the 'cygpath' conversion functionality.

Many of the issues are related to multiple hacks & workarounds that I have tried in order to overcome previous different issues, so I'd rather learn, before spending too much additional effort, if there are any incompatibilities already known between Restic's Windows build and CygWin's implementation of Bash.

The script manipulates a number of variables which end up being concatenated into a single one, which is then fed to restic. The entire command line is output to the screen before execution, and the following command line is what I see, immediately before its execution, in one example:

========================================================
   ... and ready to run the following command:

      D:/TJ/Software/Executa_sem_instalar/BackBlaze_Bkp_Restic/Exe/0.9.0/restic_0.9.0_windows_amd64.exe backup "C:\Program Files (x86)\Microsoft SQL Server Compact Edition" > D:/TJ/Software/Executa_sem_instalar/BackBlaze_Bkp_Restic/Control/log/Restic_b2_tj-0002_Backup_C_Program_Files_(x86)_Microsoft_SQL_Server_Compact_Edition_2018-06-09-13-44.log
========================================================

## What backend/server/service did you use to store the repository?

BackBlaze B2

## Expected behavior

Backup one Windows directory containing _spaces_ in pathname

## Actual behavior

error msg below :

unknown command "backup \"C:\\Program Files (x86)\\Microsoft SQL Server Compact Edition\"" for "restic"

NOTE: the actual error msg that I see on 'cmd.exe' screen (and have copied before pasting here) actually includes a good number of escape https://en.wikipedia.org/wiki/Backslash characters. These are actually present for me in the 'Write' section of this GitHub post window, but end up being skipped in the 'Preview' section.

## Steps to reproduce the behavior

As the current results are very specific to my environment and dependent on the latest iteration of unsuccessful 'fixing' that I have attempted, I'll rather wait for further discussion in order to identify what are exactly the defining elements for reproducing the issue.

## Do you have any idea what may have caused this?

Not being a software dev, I can only suspect of some kind of compatibility issue betweem Restic's Windows build and CygWin's implementation. For reference, I found the slightly similar https://github.com/restic/restic/issues/330 issue (not actually applicable, because I'm using 'cmd.exe' while the issue reported mentions some other different terminal environment).

## Do you have an idea how to solve the issue?

Maybe by some mutual investigation ... :)

## Did restic help you or made you happy in any way?

Indeed. I have started using _rclone_ , but restic's implementations of de-duplication and encryption are more interesting to me.

**Comment by fd0** [comment]
Interesting issue, I don't have any experience with that setup at all. I'm confident that your analysis is correct, and there's something odd with the quoting. The error message indicates that restic only got the following single large string as an argument:
```
backup "C:\Program Files (x86)\Microsoft SQL Server Compact Edition"
```

Since that's not a valid command, it responds with `unknown command`.

You can compile restic with debug support (`go run build.go -tags debug`), write a debug log, and look at the first line: There you'll have the arguments as restic gets them, e.g.:

```
2018/06/09 21:16:54 restic/main.go:83   main.main       1       main []string{"./restic", "backup", "foo", "/path with spaces/bar", "work"}
```

I don't have any idea what's going on here. Good luck!

**Comment by sdrubble** [comment]
Thx fd0 !

I might be just a little dense today but I'm at a loss regarding your comment _not a valid command_ ...

In the string you quoted I was passing to restic the 'backup' option, plus its required pathname. That was my exact intention - would you pls clarify what you mean by "not a valid command"  ?

Cheers

**Comment by fd0** [comment]
When restic starts, like all other programs, it gets a list of strings as the arguments for this run. Usually, this looks like this:
 * `backup`
 * `C:\Program Files (x86)\Microsoft SQL Server Compact Edition`

The first entry in this list is the restic command that is to be executed (`backup`, `snapshots`, etc).

But in your case, restic got a list with just a single entry, like this:
 * `backup "C:\Program Files (x86)\Microsoft SQL Server Compact Edition"`

Does that make it clearer?

**Comment by sdrubble** [comment]
It's indeed clear, and I'm currently in the process of tweaking the script in order to pass TWO parms to restic, instead of a single parm which concatenates those two (plus a SPACE in the middle). My first test was borked - there's apparently some kind of variable corruption in a totally different portion of the script, so now I have two different issues to investigate.

But I'm really puzzled here, because in many years of _[unfrequent...]_ usage of CygWin Bash I recall never having to worry about the **AMOUNT** of parms to be passed to any given called program... I would frequently just concatenate any amount of parms in one string, or just a few of them - with the appropriate intervening spaces **inside** the string(s) - and programs would simply parse the string and decode whatever number of parms there were inside, based on the placement of space characters.

I tried to back this statement up with some actual evidence, and did find one - with *restic* itself !!!

I have a different script - which was tested and works - where I pass **ONE monolithic string to restic.** It happily accepts the single variable below, and **lists a snapshot** as expected:

`RSTC_PROG_ARGS="ls ${SNAP_ID} -l -v"`

There's also another similar restic case (with the same `'backup'` command that's causing the current issues), where a single directory was successfully backed up thru the usage of some other monolithic string. I found the log file, but unfortunately couldn't retrieve the corresponding coding (it was a simpler, preliminary edition of the same `'backup'` script, which has currently got much fatter with alternatives and workarounds). 

**However,** there are at least **two obvious differences** though, regarding both this successful backup case and the `'list snapshot`' I mentioned above, when compared to the current issue : **no quotes**, and **no blanks-within-pathname.**
 
I expect to further investigate this and report back.

**Comment by fd0** [comment]
What about reading the list of backup target file/dir names from a file via `--files-from`? Then you don't have to pass anything with spaces in it as command-line parameters...

**Comment by sdrubble** [comment]
Sounds promising ... I'll take a look at this possibility as soon as possible !

**Comment by fd0** [comment]
Cool, glad you figured it out. I'm closing this issue for now, you can still post comments and attach files. Cheers!

**Comment by sdrubble** [comment]
Just a quick note to **take away my previous statement about "the recent slowness of CygWin Bash"...**

I had recently spent some time helping beta-test a piece of security software (HeiLig RansomOff), which I ended up deactivating due to heaviness - but I hadn't actually _uninstalled_ it. 

Last night I somehow got the inspiration to actually uninstall the software, considering that simply deactivating a service would still leave one or more drivers from it still to be reloaded on every boot, and most likely doing some silent hooking of system calls.

After uninstalling RansomOff & rebooting, the recent strangeness and slowness of CygWin Bash disappeared (and all scripts I tested were performing MUCH faster !). It's even entirely possible (haven't actually tested yet) that the errors related to quoting and spaces-in-filenames were due to some code injection happening between `bash.exe` itself and the CygWin DLL(s) which handle Windows I/O ...

Back to Restic, then. I'm currently working on polishing TWO independent script versions of 'Restic backup'. The first one of them will retain the for-loop-plus-pathnames-in-variables approach (but it will NOT touch pathnames with spaces). Second one will read the pathnames from a file (and could evidently be used for ALL pathnames regardless of space characters). I'll give some love to both versions, and might potentially end up relying on BOTH of them for dealing with varying usage cases.

Cheers
